instrumentation/xxl-job/xxl-job-common/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/xxljob/common/XxlJobSpanNameExtractor.java

```diff
@@ -0,0 +1,30 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.javaagent.instrumentation.xxljob.common;
+
+import static io.opentelemetry.javaagent.instrumentation.xxljob.common.XxlJobConstants.SCRIPT_JOB_TYPE;
+
+import com.xxl.job.core.glue.GlueTypeEnum;
+import io.opentelemetry.instrumentation.api.incubator.semconv.code.CodeAttributesGetter;
+import io.opentelemetry.instrumentation.api.incubator.semconv.code.CodeSpanNameExtractor;
+import io.opentelemetry.instrumentation.api.instrumenter.SpanNameExtractor;
+
+class XxlJobSpanNameExtractor implements SpanNameExtractor<XxlJobProcessRequest> {
+  private final SpanNameExtractor<XxlJobProcessRequest> codeSpanNameExtractor;
+
+  XxlJobSpanNameExtractor(CodeAttributesGetter<XxlJobProcessRequest> getter) {
+    codeSpanNameExtractor = CodeSpanNameExtractor.create(getter);
+  }
+
+  @Override
+  public String extract(XxlJobProcessRequest request) {
+    GlueTypeEnum glueType = request.getGlueType();
+    if (SCRIPT_JOB_TYPE.contains(glueType.getDesc())) {
+      return glueType.getDesc() + ".ID-" + request.getJobId();
```

Comment:
> span names like `GLUE(Shell).ID-2` don't follow general span naming guidelines outlined in https://github.com/open-telemetry/opentelemetry-specification/blob/v1.26.0/specification/trace/api.md#span the id portion should be removed from the span name. `GLUE(Shell)` feels a bit weird to me, but looking at the xxl-job doc I get the feeling that this is what the xxl-job ui displays. If that is the case then I think it is fine to keep it.

It backs to [discussion](https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/10421#discussion_r1502492283) above. If we remove the id portion from spanName, it will cause no way to distinguish specific GLUE(Shell) job in system. Such as there are three Shell jobs, their jobIds are 1,2,3. Without jobId in spanName, they are same spanName in APM. it's hard for user to troubleshooting.

Reply:
@trask could you provide guidance on this. Should the job id be added to the span name (I guess it doesn't have a particularly high cardinality so perhaps a bit similar to http.route) or added as an attribute to to the span?

Reply:
@steverao can you help us understand the Job ID a bit better?

a couple of questions in particular,
* how does a user correspond a Job ID back to their workflow?
* is the Job ID stable, meaning a given workflow will always have the same job ID? or can a given workflow's job ID sometimes change over time?

Reply:
> @steverao can you help us understand the Job ID a bit better?
> 
> a couple of questions in particular,
> 
> * how does a user correspond a Job ID back to their workflow?
> * is the Job ID stable, meaning a given workflow will always have the same job ID? or can a given workflow's job ID sometimes change over time?

Sure, answers to related questions are as follows:
1. The job ID is generated by job system, when users create a job in xxl-job console, it will have a unique ID. Users can find their job by job Id in console, just like picture below:
![image](https://github.com/open-telemetry/opentelemetry-java-instrumentation/assets/25423978/fdbed7c5-1f28-4607-a537-512cb757e9ee)
2. Yes, the Job ID is unique, it's a unique key to store job content in the system database. it can't be changed.

Reply:
is it possible to use the "Job description" instead of the "Job ID"? that looks more similar to existing span names for other instrumentations

Reply:
> is it possible to use the "Job description" instead of the "Job ID"? that looks more similar to existing span names for other instrumentations

I think the "Job description" is not good choose, because it has no restrictions on the input values, so jobs of the same type may have the same description.

Reply:
> jobs of the same type may have the same description

is this common?

we could still capture job id as a span attribute (just doesn't seem to be the best span name, since span name is ideally more human readable)

Reply:
>  is this common?

yes

> we could still capture job id as a span attribute (just doesn't seem to be the best span name, since span name is ideally more human readable)

ok, I understand it, but I have an alternative span name `GLUE(Shell).{jobId}`, then we can add the relevant jobId to a non ExperimentalAttributeExtractor for script job. It's will be more friendly to users to find jobId in attribute. What do you think of this? @trask @laurit

Reply:
(sorry I haven't read the code) is the Job ID only included in span name for "GLUE" type jobs? what does the span name look like for other jobs? thanks

Reply:
> (sorry I haven't read the code) is the Job ID only included in span name for "GLUE" type jobs? what does the span name look like for other jobs? thanks

In fact, the Job Id just be used in script job, such as Shell or Python script, they may not have method name. you can refer to https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/10421#discussion_r1502492283. Other jobs type are designed for java, so I can get their className and methodName, I use `className.methodName` as span name. They look like `CustomizedGroovyHandler.execute`.